### PR TITLE
fix(lsp): fix infinite loop on vim.lsp.tagfunc

### LINKED ISSUE #27325: Calling vim.lsp.tagfunc creates an infinite loop
### Problem

After https://github.com/neovim/neovim/pull/27216, calling the `vim.lsp.tagfunc` function creates an infinite loop.

This is because `vim.lsp.tagfunc()` now calls itself:

https://github.com/neovim/neovim/blob/2e982f1aad9f1a03562b7a451d642f76b04c37cb/runtime/lua/vim/lsp.lua#L2042-L2044

### Steps to reproduce

Set `'tagfunc'` to `v:lua.vim.lsp.tagfunc` and press `<C-]>`.

### Expected behavior

`vim.lsp.tagfunc` should call the actual tagfunc implementation and jump to a tag. There should not be an infinite loop.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2269+g2e982f1aa

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

N/A

### Terminal name/version

N/A

### $TERM environment variable

N/A

### Installation

Build from source

### PR DESCRIPTION
**Pull request by wookayin**
Fixes #27325. Related to #16103.

## fix(lsp): fix infinite loop on vim.lsp.tagfunc

Problem: vim.lsp.tagfunc() causes an infinite loop.

This is a bug happened while introducing deferred loading (#27216).

Solution: Rename the private module to `vim.lsp._tagfunc`.

## test(lsp): add test cases for vim.lsp.tagfunc

Problem: There is no test case for vim.lsp.tagfunc; so CI was unable to
catch the infinite loop bug (#27325).

Solution: Add test cases for vim.lsp.tagfunc().

### CODE REVIEW

File: runtime/lua/vim/lsp.lua [hunk 1]
```diff
@@ -2040,7 +2040,7 @@ end
 ---
 ---@return table[] tags A list of matching tags
 function lsp.tagfunc(pattern, flags)
-  return vim.lsp.tagfunc(pattern, flags)
+  return vim.lsp._tagfunc(pattern, flags)
```
Comment (gpanders) [inline runtime/lua/vim/lsp.lua:2043]: The purpose of this function was originally to make calling tagfunc easier (without needing `require`), but now that that isn’t necessary, this function is just a useless wrapper. Can we keep the tagfunc module named as-is and remove this wrapper function instead?

Comment (wookayin) [reply runtime/lua/vim/lsp.lua:2043]: > this function is just a useless wrapper

That's true, I also felt the same but it was bothering me that `vim.lsp.tagfunc` is a module and function (returned by the module). The problem is `gen_vimdoc`: it does not recognize the docs moved to `vim.lsp.tagfunc` --- but I fully agree with you. Let me find a solution to make docgen happy.

Comment (wookayin) [reply runtime/lua/vim/lsp.lua:2043]: If we move docs to the `vim.lsp.tagfunc` module and remove the wrapper, a new *section* in vimdoc will be added:

```
Lua module: vim.lsp.tagfunc                                      *lsp-tagfunc*

tagfunc                                              *vim.lsp.tagfunc.tagfunc*
    Provides an interface between the built-in client and 'tagfunc'.
```

But I'm not sure we want this. (BTW: it should be overridden to `*vim.lsp.tagfunc*` ...)

Comment (gpanders) [reply runtime/lua/vim/lsp.lua:2043]: This works, though it's a bit of a hack...

```diff
diff --git a/runtime/lua/vim/lsp.lua b/runtime/lua/vim/lsp.lua
index 5fa5a1db2..29f6eb8c4 100644
--- a/runtime/lua/vim/lsp.lua
+++ b/runtime/lua/vim/lsp.lua
@@ -2040,9 +2040,12 @@ end
 ---
 ---@return table[] tags A list of matching tags
 function lsp.tagfunc(pattern, flags)
-  return vim.lsp.tagfunc(pattern, flags)
+  error('Unreachable')
 end
 
+---@nodoc
+lsp.tagfunc = nil
+
 ---Checks whether a client is stopped.
 ---
 ---@param client_id (integer)
```

Comment (wookayin) [reply runtime/lua/vim/lsp.lua:2043]: Yeah, nice idea; looks hacky :D. The useless --- but useful only for the doc --- wrapper would be better though..

Comment (MariaSolOs) [reply runtime/lua/vim/lsp.lua:2043]: (Just leaving my 2 cents here) I think that dirty hacks are neat when:
- The code documents why the hack is needed and why/how it works.
- There's `git` history that provides context information around the problem (and hack).
- (Optional but still lovely) There's a `TODO` comment/GitHub issue/something documenting that this is a hack and that there's future hope for a cleaner non-hacky solution that someone will be brave enough to tackle.

In this particular situation, I believe the hack is worth it but I insist in the conditions above ;)

Comment (lewis6991) [reply runtime/lua/vim/lsp.lua:2043]: We have a similar issue with `vim.treesitter.foldexpr()` where `foldexpr` is returned by `treesitter._fold`.

This issue is really disrupting my workflow, so I'm just going to merge this as is. We can accept a follow-up PR that either fixes `gen_vimdoc.py` or adds the above "hack", but for now a redundant wrapper to a private module isn't a huge problem.

**Pull request merged**